SSA/Projects/FullyHomomorphicEncryption/Basic.lean

```diff
@@ -1,41 +1,56 @@
 import Mathlib.RingTheory.Polynomial.Quotient
+import Mathlib.RingTheory.Ideal.Quotient
 import Mathlib.Data.Zmod.Basic
 import SSA.Core.WellTypedFramework
+-- https://github.com/google/heir/blob/a18d4e8ddc0031e2a0e6dd2dd0d7fe289b9d1651/include/Dialect/Poly/IR/PolyDialect.td
+-- https://github.com/j2kun/heir/blob/5c5c0e2ff2ae37a7d1ec5791ec6c38046c4115c1/include/Dialect/Poly/IR/PolyOps.td
+-- https://github.com/google/heir/blob/a18d4e8ddc0031e2a0e6dd2dd0d7fe289b9d1651/tests/poly/poly_ops.mlir
 
 open Polynomial -- for R[X] notation
 
+
 -- Basics from Junfeng Fan and Frederik Vercauteren, Somewhat Practical Fully Homomorphic Encryption
 -- https://eprint.iacr.org/2012/144
-
 variable (q t : Nat) [Fact (q > 1)] (n : Nat)
 
--- TODO: can we make this computable?
--- failed to compile definition, consider marking it as 'noncomputable' because it depends on 'Polynomial.add'', and it does not have executable code
+-- Can we make this computable?
+-- see: https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Way.20to.20recover.20computability.3F/near/322382109
+-- and :https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Groebner.20bases
 
 -- Question: Can we make something like d := 2^n work as a macro?
+--
+
 noncomputable def f : (ZMod q)[X] := X^(2^n) + 1
 abbrev R := (ZMod q)[X] ⧸ (Ideal.span {f q n})
-
 -- Coefficients of `a : R' q n` are `a\_i : Zmod q`.
 -- TODO: get this from mathlib
-def R.coeff (a : R q n) (i : Nat) : ZMod q := sorry
 
+-- Canonical epimorphism `ZMod q[X] ->*+ R q n`
+abbrev R.fromPoly {q n : Nat} : (ZMod q)[X] → R q n := Ideal.Quotient.mk (Ideal.span {f q n})
 
+noncomputable abbrev R.zero {q n : Nat} : R q n := R.fromPoly 0
+noncomputable abbrev R.one {q n : Nat} : R q n := R.fromPoly 1
+
+noncomputable instance {q n : Nat} : Zero (R q n) := ⟨R.zero⟩
+noncomputable instance {q n : Nat} : One (R q n) := ⟨R.one⟩
+
+axiom R.representative {q n} : R q n → (ZMod q)[X]
+--axiom R.canonicalRep_fromPoly_eq {q n} : forall a : R q n, (R.fromPoly (R.representative a)) = a
 
--- TODO: get infinity norm from mathlib
---
---
 
-namespace FV
-namespace SH
+noncomputable def R.coeff (a : R q n) (i : Nat) : ZMod q := 
```

Comment:
Maybe a comment saying that this is non-canonical would be nice.  In general, I would like to aim for having readable docs from `doc-gen4`. It polishes our code, and makes it nice to upload a supplementary material of the generated documentation with the paper for reviewers to view if they like.

Reply:
fair enough! will do!